spring-shell-core/src/main/java/org/springframework/shell/command/CommandCatalog.java

```diff
@@ -0,0 +1,189 @@
+
+	/**
+	 * Interface to resolve currently existing commands. It is useful to have fully
+	 * dynamic set of commands which may exists only if some conditions in a running
+	 * shell are met. For example if shell is targeting arbitrary server environment
+	 * some commands may or may not exist depending on a runtime state.
+	 */
+	@FunctionalInterface
+	interface CommandResolver {
+
+		/**
+		 * Resolve a map of commands.
+		 *
+		 * @return a map of commands
+		 */
+		Map<String, CommandRegistration> resolve();
+	}
+
+	/**
+	 * Interface to customize a {@link CommandCatalog}.
+	 */
+	@FunctionalInterface
+	interface CommandCatalogCustomizer {
+
+		/**
+		 * Customize a command catalog.
+		 *
+		 * @param commandCatalog a command catalog
+		 */
+		void customize(CommandCatalog commandCatalog);
+	}
+
+	/**
+	 * Gets an instance of a default {@link CommandCatalog}.
+	 *
+	 * @return default command catalog
+	 */
+	static CommandCatalog of() {
+		return new DefaultCommandCatalog(null, null);
+	}
+
+	/**
+	 * Gets an instance of a default {@link CommandCatalog}.
+	 *
+	 * @param resolvers the command resolvers
+	 * @param shellContext the shell context
+	 * @return default command catalog
+	 */
+	static CommandCatalog of(Collection<CommandResolver> resolvers, ShellContext shellContext) {
+		return new DefaultCommandCatalog(resolvers, shellContext);
+	}
+
+	/**
+	 * Default implementation of a {@link CommandCatalog}.
+	 */
+	static class DefaultCommandCatalog implements CommandCatalog {
+
+		private final HashMap<String, CommandRegistration> commandRegistrations = new HashMap<>();
+		private final Collection<CommandResolver> resolvers = new ArrayList<>();
+		private final ShellContext shellContext;
+
+		DefaultCommandCatalog(Collection<CommandResolver> resolvers, ShellContext shellContext) {
+			this.shellContext = shellContext;
+			if (resolvers != null) {
+				this.resolvers.addAll(resolvers);
+			}
+		}
+
+		@Override
+		public void register(CommandRegistration... registration) {
+			for (CommandRegistration r : registration) {
+				String commandName = commandName(r.getCommands());
+				commandRegistrations.put(commandName, r);
+			}
+		}
+
+		@Override
+		public void unregister(CommandRegistration... registration) {
+			for (CommandRegistration r : registration) {
+				String commandName = commandName(r.getCommands());
+				commandRegistrations.remove(commandName);
+			}
+		}
+
+		@Override
+		public void unregister(String... commandName) {
+			for (String n : commandName) {
+				commandRegistrations.remove(n);
+			}
+		}
+
+		@Override
+		public Map<String, CommandRegistration> getRegistrations() {
+			HashMap<String, CommandRegistration> regs = new HashMap<>();
+			regs.putAll(commandRegistrations);
+			for (CommandResolver resolver : resolvers) {
+				regs.putAll(resolver.resolve());
+			}
+			return regs.entrySet().stream()
+				.filter(e -> {
+					InteractionMode mim = e.getValue().getInteractionMode();
```

Comment:
At 1st glance, this hurts my head. I get it now, but it took too long. I have found in "big" filters like this creating a well-named private method and filtering by method reference makes it more readable.

Reply:
ack